### PR TITLE
scheduling, unscheduling, and unpublishing

### PR DESCRIPTION
**Pull request by yoshokatana**
### Trello Tickets

* [schedule publish](https://trello.com/c/m9pCsSAE/211-schedule-publish)
* [unpublish](https://trello.com/c/c3VKTKGb/212-unpublish)
* [toolbar updates](https://trello.com/c/q6UNZn5j/204-l-toolbar-update)
* [publish updates](https://trello.com/c/o6hA90CF/216-publish)

### Cool New Stuff

* added ability to _unpublish_
* added ability to _schedule publish_
* added ability to _unschedule publish_
* scheduling changes publish button color in the toolbar
* publish pane is created dynamically based on publish/schedule state
* when loading the page, it will display scheduled or published status
* when _re-scheduling_, _publishing now_, or _unpublishing_, it will unschedule (if the page is scheduled to publish already) (CQ _not_ doing this was a pain point, according to editors)
* scheduling form uses native date and time fields
* added `db.urlToUri` helper method
* added `expectBooleanResult` and `db.getHead`
* added `edit.unpublishPage` to match `edit.publishPage`
* added `edit.schedulePublish` and `edit.unschedulePublish`
* added `page-state` service
* :cool: `state.formatTime` uses `moment.js`'s neat formatting rules
* added `schedule` progress color
* updated disabled button styles

### Next Steps

* add message in the publish pane giving a more detailed article state (e.g. "this was published on Friday at 10am")
* update `edit.unschedulePublish` when amphora changes (so we don't have to search through all the scheduled entries)

### CODE REVIEW

File: controllers/publish-pane.js [hunk 1]
```diff
@@ -23,18 +34,95 @@ module.exports = function () {
...
+            return edit.publishPage()
+              .then(function (url) {
+                progress.done();
+                progress.open('publish', `Published! <a href="${url}" target="_blank">View Article</a>`);
+                state.toggleScheduled(false);
+              })
+              .catch(function () {
+                // note: the Error passed into this doesn't have a message, so we use a custom one
+                progress.done('error');
+                progress.open('error', `A server error occured. Please try again.`, true);
```
Comment (gloddy) [inline controllers/publish-pane.js]: Perhaps error messages should say what was trying to be accomplished?
"Failed to Unschedule. Server error. Please try again."

File: services/pane.js [hunk 2]
```diff
@@ -55,21 +57,58 @@ function open(header, innerEl) {
...
-    `), // todo: add other publishing actions, depending on page state
-    // e.g. schedule publish, message, cancel schedule, unpublish
-    el = open(header, actionsEl);
-
-  // init controller for publish pane
-  ds.controller('publish-pane', publishPaneController);
-  ds.get('publish-pane', el.querySelector('.actions'));
+    actions = dom.create(`<div class="actions"></div>`),
+    schedule = dom.create(`<form class="schedule">
+      <input class="schedule-input" type="date" min="${today()}" value="${today()}"></input>
```
Comment (gloddy) [inline services/pane.js]: Why call today() multiple times? Why not eliminate the now() and today() functions and just replace them with vars at the top of the function?
var now = moment().format('HH:mm'),
    today = moment().format('YYYY-MM-DD');

File: services/page-state.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+ * update publish button when scheduled
+ * @param {boolean} isScheduled
+ */
+function toggleScheduled(isScheduled) {
+  var el = dom.find('.kiln-toolbar-inner .publish');
+
+  if (isScheduled) {
+    el.classList.add('scheduled');
+  } else {
+    el.classList.remove('scheduled');
```
Comment (cruzanmo) [inline services/page-state.js]: Could hold 'scheduled' string in a var.

**Comment by yoshokatana** [comment]
:+1: from morgan, christian

**Pull request merged**